kernel/src/syscall/ppoll.rs

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::{sync::atomic::Ordering, time::Duration};
+
+use super::{poll::do_sys_poll, SyscallReturn};
+use crate::{prelude::*, process::signal::sig_mask::SigMask, time::timespec_t};
+
+pub fn sys_ppoll(
+    fds: Vaddr,
+    nfds: u32,
+    timespec_addr: Vaddr,
+    sigmask_addr: Vaddr,
+    sigmask_size: usize,
+    ctx: &Context,
+) -> Result<SyscallReturn> {
+    let user_space = ctx.user_space();
+
+    let timeout = if timespec_addr != 0 {
+        let time_spec: timespec_t = user_space.read_val(timespec_addr)?;
+        Some(Duration::try_from(time_spec)?)
+    } else {
+        None
+    };
+
+    let old_sigmask = if sigmask_addr != 0 {
```

Comment:
Consider to use `signal::with_signal_blocked`.

Looks like `sys_pselect6` should use it as well.

Reply:
The logic in pselect and ppoll is not the same as `signal::with_signal_blocked`. `with_signal_blocked` will block additional signals, but here for pselect and ppoll, it just replace old sigmask with new mask.

From [the manual](https://man7.org/linux/man-pages/man2/pselect6.2.html):
> sigmask is a pointer to a signal mask (see [sigprocmask(2)](https://man7.org/linux/man-pages/man2/sigprocmask.2.html)); if it
       is not NULL, then pselect() first replaces the current signal mask
       by the one pointed to by sigmask, then does the "select" function,
       and then restores the original signal mask.  (If sigmask is NULL,
       the signal mask is not modified during the pselect() call.)

I have introduced a new method `with_sigmask_changed` for this specific behavior.

Reply:
Maybe just:
```
pub fn with_sigmask_changed<R>(
    ctx: &Context,
    mask_op: impl FnOnce(SigMask) -> SigMask,
    operate: impl FnOnce() -> R,
)
```

Then retire the `with_signal_blocked` method?